File: stronghold/code/c++/RNA.cpp
#include <iostream>
#include <string>
#include <fstream>

class RNASolution
{
private:
    std::string dna_seq, rna_seq;

public:
    RNASolution()
    {
        read_dna_seq();
        transcribing_dna_to_rna();
    }

    // lendo a sequência de DNA
    void read_dna_seq()
    {
        std::ifstream input_data("../../input/rosalind_rna.txt");

        if (input_data.is_open())
        {
            std::getline(input_data, dna_seq, '\n');

            input_data.close();
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }

    // transcrição da sequência de DNA para RNA
    void transcribing_dna_to_rna()
    {
        for (auto base : dna_seq)
        {
            if (base == 'T')
            {
                rna_seq += 'U';
                continue;
            }
            rna_seq += base;
        }
    }

    // escrevendo a sequência de rna
    void get_rna_seq()
    {
        std::ofstream output_data("../../output/RNA.txt");

        if (output_data.is_open())
        {
            output_data << rna_seq;

            output_data.close();
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo. " << std::endl;
        }
    }
};

int main()
{
    RNASolution solution;
    solution.get_rna_seq();

    return 0;
}

File: stronghold/code/python/RNA.py
import os


class RNASolution:
    def __init__(self):
        self.dna = None
        self.read_dna_seq()

    # lendo a sequência do dna
    def read_dna_seq(self):
        try:
            path = os.path.realpath("../../input/rosalind_rna.txt")
            input_data = open(path, 'r')
            self.dna = input_data.read().strip()
            input_data.close()

        except IOError:
            print("Erro na leitura do arquivo!")

    # transcrevendo a sequência de dna para rna
    def transcribing_dna_to_rna(self):
        rna = ""

        for base in self.dna:
            if base == "T":
                rna += "U"
                continue
            rna += base

        return rna

    # sequência rna
    def getRnaSeq(self):
        rna = self.transcribing_dna_to_rna()

        try:
            path = os.path.realpath("../../output/RNA.txt")
            output_data = open(path, 'w')
            output_data.write(rna)
            output_data.close()

        except IOError:
            print("Erro na escrita do arquivo!")


if __name__ == "__main__":
    solution = RNASolution()
    solution.getRnaSeq()
